### ISSUE TITLE
Framework doesn't work with images that are not 640x480

### ISSUE THREAD

**Issue by mtmarsh2** [Russell91/TensorBox#6]
Hi, I'm having trouble training on a subset of Coco, and when I try the evaluate.ipynb code, no boxes are shown on the images until I reduce the min_accuracy to a very small number, and then each example image is filled with a grid of boxes, which means the lstms are submitting every box.

I tried debugging this by setting the train idl file to images of dogs, and the test idl file to images of clocks, but while training, the Test Accuracy is still outputting near 99%, which doesn't make sense as I'm purposely trailing on one class and testing on a totally different class. How exactly is the Test Accuracy computed?

**Comment by Russell91** [comment]
Okay, one explanation would be that just no boxes are getting loaded, and the model is outputting 0 confidence for everything (very accurate if nothing is there). Can you please try loading tensorboard and paste a screenshot of the Images tab? It should let you verify that the ground truth boxes are indeed being encoded.

**Comment by mtmarsh2** [comment]
It seems like the boxes are getting loaded, as each of the truth images has the bounding box in green, but the pred images don't have any bounding boxes: 
<img width="1191" alt="screen shot 2016-02-27 at 7 20 45 pm" src="https://cloud.githubusercontent.com/assets/2382066/13376664/616a86a2-dd87-11e5-9f83-076acd483aae.png">

**Comment by Russell91** [comment]
Can you try setting the training and test sets to the same idl file with just 1 image, that indeed contains a clock? The tensorboard images debugging screen should be more useful here than the evaluate.ipynb.

**Comment by mtmarsh2** [comment]
I tried that, I took one picture of a clock and put it in both idl files, the accuracy maxed at 100% and when I looked at the "images" tab in tensorboard, the predicted images all had boxes drawn on them. I then tried it with a set of 5 images for clocks, then 15 for clocks, and both had prediction boxes drawn in the correct areas. Next I tried 50, and even though the accuracy is topping at 90%, there are no prediction boxes drawn on the predicted images, which is weird. But it looks like the loss is fluzuating alot, maybe thats the problem? 
<img width="602" alt="screen shot 2016-02-27 at 11 22 48 pm" src="https://cloud.githubusercontent.com/assets/2382066/13377651/78867d60-dda9-11e5-8936-51a6f05187b9.png">

**Comment by Russell91** [comment]
Can you confirm that you're using the latest version of the repo and the default hyperparam files? The regression loss also seems much bigger than it should be.

**Comment by mtmarsh2** [comment]
Yep just downloaded the repo again and reran, same results with the clock class. I tried running on 50 images of the dog class in coco, and through the images tab of tensor board, it seems like the images where the dog is center have prediction boxes drawn, yet the images where the dog is very small none are drawn. Maybe it's just that the data is too varied, vs the face dataset has mostly similar size faces?

**Comment by Russell91** [comment]
Having highly varied data is indeed a problem for these end-to-end models. That's why Faster-RCNN uses 9 separate decoders across 3 different scales and 3 different aspect ratios. If you're running on CoCo, you might be better off without the lstm decoder as well, as it's unlike to add anything other than complexity when there are no boxes in close proximity.

**Comment by Russell91** [comment]
Also, you may just not be training for long enough. It can take 10k iterations for boxes to show up. It occurred to me why the small boxes are not showing up though - the confidence maps test whether an object intersects a cell of the 15x20 grid over the image, so bigger boxes intersect more cells, and hence learn faster. One trick could be to discount the loss of larger boxes and increase the overall learning rate.

**Comment by mtmarsh2** [comment]
Interesting thanks for the pointer, I'll try using a subset of the category within Coco and limit the instances to having about the same area, so they are of the same scale. I tried 14k iterations on 2000 instances in Coco, and the Test accuracy reported in the console is consistently around 85%, yet when I click on the images tab in tensorboard, no predictions are drawn. Why would the accuracy be so high if no predictions are drawn, wouldn't it be low according to IoU?

**Comment by Russell91** [comment]
Boxes are only drawn when the confidence is greater than 0.5. You can lower that threshold in train.py and see if it helps draw boxes sooner. The accuracy is not measuring IoU, but whether the confidence is guessed correctly. In this case, 85% just means guessing 0 every time. You're better off watching progress on the train confidence_loss than the test accuracy early on.

**Comment by mtmarsh2** [comment]
Cool I'll try that, thanks! What would line would I change for that threshold in train.py? I found the threshold in "stich_wrapper" which renders the boxes, is that the one you were talking about?

**Comment by Russell91** [comment]
I will also suggest that if you're running on images with larger boxes, you try getting the default.json model to work first. I'm seeing earlier success with that model. The lstm may end up requiring a lower learning rate or other tweaks on the bigbox datasets. The caffe repo used gradient clipping, which might explain the better performance of a single hyperparameter setting across datasets.

**Comment by Russell91** [comment]
Okay, yea. I'm seeing the same terrible regression loss that you are on some of these datasets with the lstm model. But reducing the learning rate by a factor of 10 fixes the issue and I'm getting decent box predictions after just 1k iterations.

**Comment by mtmarsh2** [comment]
I tried lowering the learning rate to .0001 for the default/hypes.json file, but the problem seems to be that the net stops predicting boxes after a short time. In the screenshot, the left column was a job I ran for ~30 seconds, and you can see that the net predicted a bunch of boxes. The job to the right I ran for 2000 iterations, and no boxes appeared at all as I checked the page. I verified this afterwards by running a new test and immediately saw boxes, but after a minute the net stopped predicting. Not sure why
<img width="317" alt="screen shot 2016-03-03 at 1 09 12 pm" src="https://cloud.githubusercontent.com/assets/2382066/13506022/95f8b35e-e141-11e5-8961-f1bcae5f4c33.png">

**Comment by Russell91** [comment]
Are you seeing these issues for the default.json as well? That should be more robust.

**Comment by mtmarsh2** [comment]
Weird, I reinstalled the repo and tried lowering the learning rate again and the results were very good, thanks again for the help!

**Comment by Russell91** [comment]
Okay, great! I've gone ahead and reduced the default learning rate for the lstm to 0.0003. That seems to be working better for me on the larger box datasets.

**Comment by Russell91** [comment]
Okay guys, sorry to have failed you here. The issue is that while the framework "works" for images that have to be resized to the values in the .json file, the resize function only works for the first epoch. On the second epoch, the boxes are resized again. And again for the third epoch... Until all boxes are either hugely massive or 0. Anyways, I'm going to try and push a fix for this.

**Comment by Russell91** [comment]
Okay, this has been addressed in master by https://github.com/Russell91/TensorBox/pull/12.

**Comment by Russell91** [comment]
I've again reopened this issue temporarily. Embarrassingly, I forgot to merge dev into master after fixing this bug, and thus the changes didn't propagate.
